<a href="https://colab.research.google.com/github/ami-null/python-course/blob/main/12_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas

`pandas` is an open-source module for Python that contains data structures and data manipulation tools designed to make data cleaning and analysis fast and convenient in Python.

It adopts significant parts of NumPy's idiomatic style of array-based computing, especially array-based functions and a preference for data processing without for loops.

While `pandas` adopts many coding idioms from `NumPy`, the biggest difference is that `pandas` is designed for working with tabular or heterogeneous data. `NumPy`, by contrast, is best suited for working with homogeneously typed numerical array of one or more dimensions.

## Learning Objectives

By the end of this lecture, you should be able to:

- Create and using `pandas` `Series` and `DataFrame` objects
- Load data from a file or URL
- Identify and handle missing values with `.isnull()`, `.dropna()`, and `.fillna()`
- Select rows and columns using `.iloc` and boolean filtering
- Sort a `DataFrame` and summarize it with `.describe()`, `.value_counts()`, and `.corr()`
- Create new columns with `.transform()` and `.assign()`
- Discretize continuous variables using `pd.cut()` and `pd.qcut()`
- Summarize data with `.agg()` across several equivalent syntaxes
- Apply custom functions across rows or columns with `.apply()`
- Perform grouped aggregation, transformation, and mutation with `.groupby()`
- Combine two `DataFrame`s with `.merge()`

## Installation

Run the following in the terminal:
`pip install pandas`

If running inside Jupyter/Colab notebook, run:
`!pip install pandas`

## Getting Started

Two central concepts in `pandas` are: `Series` and `DataFrame`.

- A `Series` is like a vector or a column containing some values of the same datatype
- A `DataFrame` is a table consisting of one or more `Series`. Each series can have its own datatype

## `Series`
- A Series is a one-dimensional array-like object containing a sequence of values (of similar types to NumPy types) of the same type
- Every `Series` has an associated array of data labels, called its `index`

In [ ]:
import pandas as pd

In [ ]:
obj = pd.Series([4, 7, -5, 3])
obj

0    4
1    7
2   -5
3    3
dtype: int64

In [ ]:
obj[0:2]    # we can perform slicing just like in numpy

0    4
1    7
dtype: int64

Since we did not specify an index for the data, a default one consisting of the integers 0 through N-1 (where N is the length of the data) is created

We can also create a Series with an index identifying each data point with a label:

In [ ]:
obj2 = pd.Series([4, 7, -5, 3], index=["d", "b", "a", "c"])
obj2

d    4
b    7
a   -5
c    3
dtype: int64

In [ ]:
obj2["b"]

np.int64(7)

## `DataFrame`
A `DataFrame` represents a rectangular table of data and contains an ordered, named collection of columns, each of which can be a different value type (numeric, string, Boolean, etc.).

There are many ways to construct a `DataFrame` from scratch, though one of the most common is from a dictionary (`dict`) of equal-length `list`s or `NumPy` arrays

In [ ]:
data = {"state": ["Ohio", "Ohio", "Ohio", "Nevada", "Nevada", "Nevada"],
        "year": [2000, 2001, 2002, 2001, 2002, 2003],
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
frame = pd.DataFrame(data)
frame

,state,year,pop
0,Ohio,2000,1.5
1,Ohio,2001,1.7
2,Ohio,2002,3.6
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


The resulting `DataFrame` has its `index` assigned automatically, as with `Series`, and the columns are placed according to the order of the keys in data (which depends on their insertion order in the dictionary)

## Reading and Writing Data using `pandas`

| Function       | Description                                                                                              |
|----------------|----------------------------------------------------------------------------------------------------------|
| read_csv       | Load delimited data from a file, URL, or file-like object; use comma as default delimiter                |
| read_clipboard | Variation of read_csv that reads data from the clipboard; useful for converting tables from web pages    |
| read_excel     | Read tabular data from an Excel XLS or XLSX file                                                         |
| read_json      | Read data from a JSON (JavaScript Object Notation) string representation, file, URL, or file-like object |
| read_sas       | Read a SAS dataset stored in one of the SAS system's custom storage formats                              |
| read_spss      | Read a data file created by SPSS                                                                         |
| read_stata     | Read a dataset from Stata file format                                                                    |


## Loading the data

We shall now load the IMDb Top 10000 Movies dataset (https://www.kaggle.com/datasets/moazeldsokyx/imdb-top-10000-movies-dataset). It contains rating, runtime, genre, metascore, plot, directors, stars, votes, gross earnings etc of 9999 movies up to June 2023.

In [ ]:
movies_df = pd.read_csv("https://github.com/ami-null/python-course/raw/refs/heads/main/data/Top_10000_Movies_IMDb.csv")

After loading the data, we should view the basic informations regarding the columns:

In [ ]:
movies_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          9999 non-null   int64  
 1   Movie Name  9999 non-null   str    
 2   Rating      9999 non-null   float64
 3   Runtime     9999 non-null   str    
 4   Genre       9999 non-null   str    
 5   Metascore   7992 non-null   float64
 6   Plot        9999 non-null   str    
 7   Directors   9999 non-null   str    
 8   Stars       9999 non-null   str    
 9   Votes       9999 non-null   int64  
 10  Gross       9999 non-null   int64  
 11  Link        9999 non-null   str    
dtypes: float64(2), int64(3), str(7)
memory usage: 937.5 KB


In [ ]:
movies_df.shape

(9999, 12)

We see that the data has 12 columns, 9999 rows, and only the `Metascore` column has missing values. It also shows the data type of each column. The columns with `Dtype` as `object` can be strings, lists etc.

We should also print out a few rows of the dataset to keep as a visual reference. We accomplish this with `.head()` method:

In [ ]:
movies_df.head()

,ID,Movie Name,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.0,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",NaN,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",NaN,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/
4,5,Daman,9.1,121 min,"Adventure, Drama",NaN,"The film is set in 2015. Sid, is a young docto...","['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303,https://www.imdb.com/title/tt17592606/


`.head()` outputs the first five rows of a DataFrame by default, but we could also pass a number as well: `movies_df.head(10)` would output the first ten rows, for example.

Further, in the above output, we see that the `Metascore` column has some `NaN` values. `NaN` means "not a number" and indicates missing values in the case of numeric columns. In columns with other data types, `None` or `<NA>` might be used to indicated missing values.

## Dropping unnecessary columns

Often we do not need to keep all the columns in dataset. Dropping the `Plot` and the `Link` column:

In [ ]:
movies_df = movies_df.drop(columns= ['Plot', 'Link'])
movies_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          9999 non-null   int64  
 1   Movie Name  9999 non-null   str    
 2   Rating      9999 non-null   float64
 3   Runtime     9999 non-null   str    
 4   Genre       9999 non-null   str    
 5   Metascore   7992 non-null   float64
 6   Directors   9999 non-null   str    
 7   Stars       9999 non-null   str    
 8   Votes       9999 non-null   int64  
 9   Gross       9999 non-null   int64  
dtypes: float64(2), int64(3), str(5)
memory usage: 781.3 KB


In [ ]:
# we don't want spaces inside column names
movies_df = movies_df.rename(columns = {"Movie Name" : "title"})

To see the last five rows, we can use `.tail()` method.

`.tail()` also accepts a number, and in this case we printing the bottom three rows:

In [ ]:
movies_df.tail(3)

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
9996,9997,Paranormal Activity 4,4.6,88 min,"Horror, Mystery",40.0,"['Henry Joost', 'Ariel Schulman', 'Stephen Dun...","['Ariel Schulman', 'Stephen Dunham', 'Katie Fe...",68430,53921300
9997,9998,Dark Crimes,4.6,92 min,"Crime, Drama, Mystery",24.0,"['Alexandros Avranas', 'Jim Carrey', 'Marton C...","['Jim Carrey', 'Marton Csokas', 'Charlotte Gai...",19098,19098
9998,9999,The Pyramid,4.6,89 min,"Horror, Mystery, Thriller",24.0,"['Grégory Levasseur', 'Ashley Hinshaw', 'James...","['Ashley Hinshaw', 'James Buckley', ""Denis O'H...",25483,2756333


## Handling missing values

First, we have check which columns have missing values. The `.isnull()` method returns a `DataFrame` containing the same number of rows and columns as the original `DataFrame` but each cell is either `True` or `False` indicating whether the data has missing value in the corresponding cell.

In [ ]:
movies_df.isnull()

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
0,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,True,False,False,False,False
3,False,False,False,False,False,True,False,False,False,False
4,False,False,False,False,False,True,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...
9994,False,False,False,False,False,False,False,False,False,False
9995,False,False,False,False,False,False,False,False,False,False
9996,False,False,False,False,False,False,False,False,False,False
9997,False,False,False,False,False,False,False,False,False,False


In [ ]:
# we can sum the output of .isnull() to get count of missing values
movies_df.isnull().sum()

ID              0
title           0
Rating          0
Runtime         0
Genre           0
Metascore    2007
Directors       0
Stars           0
Votes           0
Gross           0
dtype: int64

The `Metascore` column has some missing values.

### Complete case analysis

We can drop the missing rows containing values and perform further analysis with only the complete observations (this is known as complete case analysis).

In [ ]:
movies_cc = movies_df.dropna()
movies_cc.isna().sum()    # the .isna() method can also be used

ID           0
title        0
Rating       0
Runtime      0
Genre        0
Metascore    0
Directors    0
Stars        0
Votes        0
Gross        0
dtype: int64

### Imputation
Imputation means filling out the missing values in a dataset with plausible values based on some criteria/logic/algorithm. There are many basic and advanced forms of imputation, however, they are out of the scope of this course. Therefore, we shall limit discussion on imputation at a minimum.

Missing values can be filled with the `.fillna()` method.

In [ ]:
# replacing the missing values with the overall average value
movies_df['Metascore'] = movies_df['Metascore'].fillna(movies_df['Metascore'].mean())

In [ ]:
movies_df['Metascore'].isnull().sum()    # missing count 0 now

np.int64(0)

## Indexing and Slicing a `DataFrame`

### Selecting a single column

In [ ]:
genre_col = movies_df['Genre']

In [ ]:
genre_col

0                              Drama
1                       Crime, Drama
2       Animation, Action, Adventure
3                      Comedy, Drama
4                   Adventure, Drama
                    ...             
9994    Animation, Adventure, Comedy
9995         Comedy, Fantasy, Horror
9996                 Horror, Mystery
9997           Crime, Drama, Mystery
9998       Horror, Mystery, Thriller
Name: Genre, Length: 9999, dtype: str

In [ ]:
print(type(genre_col))

<class 'pandas.Series'>


In [ ]:
print(type(movies_df.Genre))    # same as above

<class 'pandas.Series'>


We can see above that selecting a single column with the `df['col_name']` syntax or the `df.col_name` syntax returns a `Series` object.

### Selecting multiple columns

We can also select one or more columns with `df[['col1', 'col2',...]` syntax. In this case the return value would of type `DataFrame` (even if we select only one column)

In [ ]:
movies_df[['title', 'Runtime']].head(4)

,title,Runtime
0,The Shawshank Redemption,142 min
1,The Godfather,175 min
2,Ramayana: The Legend of Prince Rama,135 min
3,The Chaos Class,87 min


In [ ]:
print(type(movies_df[['title', 'Runtime']]))

<class 'pandas.DataFrame'>


In [ ]:
print(type(movies_df[['title']]))

<class 'pandas.DataFrame'>


### Selecting Columns Based on Data Types

Below we select the numeric columns:

In [ ]:
movies_num = movies_df.select_dtypes("number")
movies_num.info()

<class 'pandas.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ID         9999 non-null   int64  
 1   Rating     9999 non-null   float64
 2   Metascore  9999 non-null   float64
 3   Votes      9999 non-null   int64  
 4   Gross      9999 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 390.7 KB


### Selecting Rows

`pandas` offers two accessors for selecting rows: `.loc`, which locates by label, and `.iloc`, which locates by integer position. This lecture covers the basics of `.iloc`.

#### Using `.iloc`

In [ ]:
movies_df.iloc[0:4]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.000000,"['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785


In `.iloc` we can also mention specific columns to select. Inside the brackets, the indices before the comma specifies rows and indices after the comma specifies columns:

In [ ]:
movies_df.iloc[0:4, 1:3]    # First four rows, 2nd and 3rd columns

,title,Rating
0,The Shawshank Redemption,9.3
1,The Godfather,9.2
2,Ramayana: The Legend of Prince Rama,9.2
3,The Chaos Class,9.2


## Filtering Values (selecting rows based on conditions)

Selecting movies with a high rating:

In [ ]:
movies_df[movies_df['Rating'] > 8]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.000000,"['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303
...,...,...,...,...,...,...,...,...,...,...
491,492,Anand,8.1,122 min,"Drama, Musical",58.749625,"['Hrishikesh Mukherjee', 'Rajesh Khanna', 'Ami...","['Rajesh Khanna', 'Amitabh Bachchan', 'Sumita ...",34462,34462
492,493,Iqbal,8.1,132 min,"Drama, Sport",58.749625,"['Nagesh Kukunoor', 'Shreyas Talpade', 'Naseer...","['Shreyas Talpade', 'Naseeruddin Shah', 'Shwet...",17329,17329
493,494,The Legend of Bhagat Singh,8.1,155 min,"Action, Biography, Drama",58.749625,"['Rajkumar Santoshi', 'Ajay Devgn', 'Sushant S...","['Ajay Devgn', 'Sushant Singh', 'D. Santosh', ...",16881,16881
494,495,Ghilli,8.1,160 min,"Action, Drama, Romance",58.749625,"['Dharani', 'Joseph Vijay', 'Trisha Krishnan',...","['Joseph Vijay', 'Trisha Krishnan', 'Prakash R...",15531,15531


Sometimes a high rating alone may not be indicative of a good movie, so we can filter with a high number of votes too:

In [ ]:
movies_df[(movies_df['Rating'] > 8) & (movies_df['Votes'] > 100000)]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.000000,"['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818
...,...,...,...,...,...,...,...,...,...,...
408,409,Wild Strawberries,8.1,91 min,"Drama, Romance",88.000000,"['Ingmar Bergman', 'Victor Sjöström', 'Bibi An...","['Victor Sjöström', 'Bibi Andersson', 'Ingrid ...",110195,110195
417,418,Lagaan: Once Upon a Time in India,8.1,224 min,"Drama, Musical, Sport",84.000000,"['Ashutosh Gowariker', 'Aamir Khan', 'Raghubir...","['Aamir Khan', 'Raghubir Yadav', 'Gracy Singh'...",117077,70147
419,420,Mr. Smith Goes to Washington,8.1,129 min,"Comedy, Drama",73.000000,"['Frank Capra', 'James Stewart', 'Jean Arthur'...","['James Stewart', 'Jean Arthur', 'Claude Rains...",117774,9600000
432,433,The Gold Rush,8.1,95 min,"Adventure, Comedy, Drama",58.749625,"['Charles Chaplin', 'Charles Chaplin', 'Mack S...","['Charles Chaplin', 'Mack Swain', 'Tom Murray'...",114081,5450000


Filtering movies which have only "Action" in their genre:

In [ ]:
movies_df[movies_df["Genre"] == "Action"]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
85,86,Banshee,8.5,50 min,Action,58.749625,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095
1684,1685,Ayan,7.6,162 min,Action,58.749625,"['K.V. Anand', 'Suriya', 'Tamannaah Bhatia', '...","['Suriya', 'Tamannaah Bhatia', 'Prabhu', 'Jagan']",11086,11086
7284,7285,Ong Bak 2,6.2,98 min,Action,47.000000,"['Tony Jaa', 'Panna Rittikrai', 'Tony Jaa', 'S...","['Panna Rittikrai', 'Tony Jaa', 'Sarunyu Wongk...",31338,102055
7707,7708,Velayudham,6.1,186 min,Action,58.749625,"['Mohan Raja', 'Joseph Vijay', ""Genelia D'Souz...","['Joseph Vijay', ""Genelia D'Souza"", 'Hansika M...",12375,12375
9238,9239,The Man with the Iron Fists,5.4,95 min,Action,51.000000,"['RZA', 'Russell Crowe', 'Cung Le', 'Lucy Liu'...","['Russell Crowe', 'Cung Le', 'Lucy Liu', 'RZA']",63551,15634090
9825,9826,Ong Bak 3,4.9,99 min,Action,38.000000,"['Tony Jaa', 'Panna Rittikrai', 'Tony Jaa', 'D...","['Panna Rittikrai', 'Tony Jaa', 'Dan Chupong',...",14331,14498


Filtering movies which have "Action" or "Comedy" in their genre:

In [ ]:
movies_df[(movies_df["Genre"] == "Action") | (movies_df["Genre"] == "Comedy")]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
14,15,Tosun Pasa,8.9,90 min,Comedy,58.749625,"['Kartal Tibet', 'Kemal Sunal', 'Müjde Ar', 'A...","['Kemal Sunal', 'Müjde Ar', 'Adile Nasit', 'Se...",23933,23933
15,16,The Chaos Class Failed the Class,8.9,91 min,Comedy,58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Semra Özdamar'...",23976,23976
85,86,Banshee,8.5,50 min,Action,58.749625,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095
140,141,"When I Grow Up, I'll Be a Kangaroo",8.4,92 min,Comedy,58.749625,"['Radivoje Andric', 'Sergej Trifunovic', 'Mari...","['Sergej Trifunovic', 'Marija Karan', 'Nebojsa...",11116,11116
520,521,Young Frankenstein,8.0,106 min,Comedy,83.000000,"['Mel Brooks', 'Gene Wilder', 'Madeline Kahn',...","['Gene Wilder', 'Madeline Kahn', 'Marty Feldma...",163092,86300000
...,...,...,...,...,...,...,...,...,...,...
9916,9917,Recep Ivedik,4.8,90 min,Comedy,58.749625,"['Togan Gökbakar', 'Sahan Gökbakar', 'Fatma To...","['Sahan Gökbakar', 'Fatma Toptas', 'Tulug Çizg...",27759,27759
9921,9922,The Bubble,4.7,126 min,Comedy,34.000000,"['Judd Apatow', 'Harry Trevaldwyn', 'Samson Ka...","['Harry Trevaldwyn', 'Samson Kayo', 'Peter Ser...",26922,26922
9954,9955,Deuce Bigalow: European Gigolo,4.7,83 min,Comedy,23.000000,"['Mike Bigelow', 'Rob Schneider', 'Eddie Griff...","['Rob Schneider', 'Eddie Griffin', 'Jeroen Kra...",52178,22264487
9961,9962,Like a Boss,4.7,83 min,Comedy,33.000000,"['Miguel Arteta', 'Tiffany Haddish', 'Rose Byr...","['Tiffany Haddish', 'Rose Byrne', 'Salma Hayek...",14781,14781


The above query can be concisely written using the `.isin()` method:

In [ ]:
movies_df[movies_df["Genre"].isin(["Action", "Comedy"])]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
14,15,Tosun Pasa,8.9,90 min,Comedy,58.749625,"['Kartal Tibet', 'Kemal Sunal', 'Müjde Ar', 'A...","['Kemal Sunal', 'Müjde Ar', 'Adile Nasit', 'Se...",23933,23933
15,16,The Chaos Class Failed the Class,8.9,91 min,Comedy,58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Semra Özdamar'...",23976,23976
85,86,Banshee,8.5,50 min,Action,58.749625,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095
140,141,"When I Grow Up, I'll Be a Kangaroo",8.4,92 min,Comedy,58.749625,"['Radivoje Andric', 'Sergej Trifunovic', 'Mari...","['Sergej Trifunovic', 'Marija Karan', 'Nebojsa...",11116,11116
520,521,Young Frankenstein,8.0,106 min,Comedy,83.000000,"['Mel Brooks', 'Gene Wilder', 'Madeline Kahn',...","['Gene Wilder', 'Madeline Kahn', 'Marty Feldma...",163092,86300000
...,...,...,...,...,...,...,...,...,...,...
9916,9917,Recep Ivedik,4.8,90 min,Comedy,58.749625,"['Togan Gökbakar', 'Sahan Gökbakar', 'Fatma To...","['Sahan Gökbakar', 'Fatma Toptas', 'Tulug Çizg...",27759,27759
9921,9922,The Bubble,4.7,126 min,Comedy,34.000000,"['Judd Apatow', 'Harry Trevaldwyn', 'Samson Ka...","['Harry Trevaldwyn', 'Samson Kayo', 'Peter Ser...",26922,26922
9954,9955,Deuce Bigalow: European Gigolo,4.7,83 min,Comedy,23.000000,"['Mike Bigelow', 'Rob Schneider', 'Eddie Griff...","['Rob Schneider', 'Eddie Griffin', 'Jeroen Kra...",52178,22264487
9961,9962,Like a Boss,4.7,83 min,Comedy,33.000000,"['Miguel Arteta', 'Tiffany Haddish', 'Rose Byr...","['Tiffany Haddish', 'Rose Byrne', 'Salma Hayek...",14781,14781


But the `Genre` column is a list and has multiple genre name for most movies. So, instead of checking equality with `"Action"`, we should instead check whether `"Action"` is contained in the `Genre` column:

In [ ]:
movies_df[movies_df["Genre"].str.contains("Action")]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444
10,11,The Lord of the Rings: The Return of the King,9.0,201 min,"Action, Adventure, Drama",94.000000,"['Peter Jackson', 'Elijah Wood', 'Viggo Morten...","['Elijah Wood', 'Viggo Mortensen', 'Ian McKell...",1890885,377845905
13,14,Demon Slayer: Kimetsu no Yaiba - Tsuzumi Mansi...,8.9,87 min,"Animation, Action, Fantasy",58.749625,['Haruo Sotozaki'],[],18738,18738
...,...,...,...,...,...,...,...,...,...,...
9981,9982,Bodyguard,4.7,130 min,"Action, Comedy, Drama",41.000000,"['Siddique', 'Salman Khan', 'Kareena Kapoor', ...","['Salman Khan', 'Kareena Kapoor', 'Raj Babbar'...",26919,26919
9984,9985,Superhero Movie,4.6,75 min,"Action, Comedy, Sci-Fi",33.000000,"['Craig Mazin', 'Drake Bell', 'Leslie Nielsen'...","['Drake Bell', 'Leslie Nielsen', 'Sara Paxton'...",72271,26638520
9986,9987,Interceptor,4.6,99 min,"Action, Adventure, Drama",51.000000,"['Matthew Reilly', 'Elsa Pataky', 'Luke Bracey...","['Elsa Pataky', 'Luke Bracey', 'Aaron Glenane'...",25686,25686
9988,9989,Aliens vs. Predator: Requiem,4.6,94 min,"Action, Adventure, Horror",29.000000,"['Colin Strause', 'Greg Strause', 'Reiko Ayles...","['Greg Strause', 'Reiko Aylesworth', 'Steven P...",129456,41797066


Selecting movies with gross revenue above the overall 75% percentile of all movies in the datset:

In [ ]:
movies_df[movies_df['Gross'] > movies_df['Gross'].quantile(0.75)]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.0,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.0,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.0,"['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.0,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000
10,11,The Lord of the Rings: The Return of the King,9.0,201 min,"Action, Adventure, Drama",94.0,"['Peter Jackson', 'Elijah Wood', 'Viggo Morten...","['Elijah Wood', 'Viggo Mortensen', 'Ian McKell...",1890885,377845905
...,...,...,...,...,...,...,...,...,...,...
9988,9989,Aliens vs. Predator: Requiem,4.6,94 min,"Action, Adventure, Horror",29.0,"['Colin Strause', 'Greg Strause', 'Reiko Ayles...","['Greg Strause', 'Reiko Aylesworth', 'Steven P...",129456,41797066
9990,9991,Hannah Montana: The Movie,4.6,102 min,"Comedy, Drama, Family",47.0,"['Peter Chelsom', 'Miley Cyrus', 'Emily Osment...","['Miley Cyrus', 'Emily Osment', 'Billy Ray Cyr...",44174,79566871
9991,9992,God's Not Dead,4.6,113 min,"Adventure, Comedy, Drama",16.0,"['Harold Cronk', 'Shane Harper', 'Kevin Sorbo'...","['Shane Harper', 'Kevin Sorbo', 'David A.R. Wh...",43934,60755732
9994,9995,Yogi Bear,4.6,80 min,"Animation, Adventure, Comedy",35.0,"['Eric Brevig', 'Dan Aykroyd', 'Justin Timberl...","['Dan Aykroyd', 'Justin Timberlake', 'Anna Far...",22762,100246011


In [ ]:
# filtering movies with rating between 8 and 9:
movies_df[(movies_df["Rating"] >= 8) & (movies_df["Rating"] <= 9)]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818
8,9,12 Angry Men,9.0,96 min,"Crime, Drama",97.000000,"['Sidney Lumet', 'Henry Fonda', 'Lee J. Cobb',...","['Henry Fonda', 'Lee J. Cobb', 'Martin Balsam'...",815205,4360000
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.000000,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000
...,...,...,...,...,...,...,...,...,...,...
670,671,My Grandfather's People,8.0,126 min,"Comedy, Drama, Family",58.749625,"['Çagan Irmak', 'Çetin Tekindor', 'Yigit Özsen...","['Çetin Tekindor', 'Yigit Özsener', 'Gökçe Bah...",11023,11023
671,672,Manjhi: The Mountain Man,8.0,120 min,"Biography, Drama",58.749625,"['Ketan Mehta', 'Nawazuddin Siddiqui', 'Abhina...","['Nawazuddin Siddiqui', 'Abhinay Raj Singh', '...",19577,19577
672,673,Pokiri,8.0,164 min,"Action, Crime, Thriller",58.749625,"['Puri Jagannadh', 'Mahesh Babu', ""Ileana D'Cr...","['Mahesh Babu', ""Ileana D'Cruz"", 'Prakash Raj'...",15030,15030
673,674,Cars of the Revolution,8.0,115 min,Drama,58.749625,"['Tolga Örnek', 'Haluk Bilginer', 'Taner Birse...","['Haluk Bilginer', 'Taner Birsel', 'Charles Ca...",12416,12416


The above can be rewritten more concisely using the `.between()` method:

In [ ]:
movies_df[movies_df["Rating"].between(8, 9)]

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818
8,9,12 Angry Men,9.0,96 min,"Crime, Drama",97.000000,"['Sidney Lumet', 'Henry Fonda', 'Lee J. Cobb',...","['Henry Fonda', 'Lee J. Cobb', 'Martin Balsam'...",815205,4360000
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.000000,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000
...,...,...,...,...,...,...,...,...,...,...
670,671,My Grandfather's People,8.0,126 min,"Comedy, Drama, Family",58.749625,"['Çagan Irmak', 'Çetin Tekindor', 'Yigit Özsen...","['Çetin Tekindor', 'Yigit Özsener', 'Gökçe Bah...",11023,11023
671,672,Manjhi: The Mountain Man,8.0,120 min,"Biography, Drama",58.749625,"['Ketan Mehta', 'Nawazuddin Siddiqui', 'Abhina...","['Nawazuddin Siddiqui', 'Abhinay Raj Singh', '...",19577,19577
672,673,Pokiri,8.0,164 min,"Action, Crime, Thriller",58.749625,"['Puri Jagannadh', 'Mahesh Babu', ""Ileana D'Cr...","['Mahesh Babu', ""Ileana D'Cruz"", 'Prakash Raj'...",15030,15030
673,674,Cars of the Revolution,8.0,115 min,Drama,58.749625,"['Tolga Örnek', 'Haluk Bilginer', 'Taner Birse...","['Haluk Bilginer', 'Taner Birsel', 'Charles Ca...",12416,12416


### A common pandas pitfall: chained indexing

When you filter a `DataFrame` and then try to assign a new value to the filtered result in the same statement, pandas cannot always tell whether you are working with a view or a copy of the original data. This produces a `SettingWithCopyWarning`, and the assignment may silently fail to update anything.

The fix is to explicitly assign the filtered result to a new variable with `.copy()`, so pandas knows you intend to work with an independent copy rather than a view of the original data.

In [ ]:
# UNCOMMENT TO SEE THE WARNING
# chained indexing: filter, then assign, in two separate bracket operations, without copying first
# high_rated = movies_df[movies_df['Rating'] > 9]
# high_rated['Genre'] = 'Top Tier'    # triggers SettingWithCopyWarning

In [ ]:
# correct approach: .copy() makes the new variable an independent DataFrame, not a view
high_rated = movies_df[movies_df['Rating'] > 9].copy()
high_rated['Genre'] = 'Top Tier'
high_rated.head()

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
0,1,The Shawshank Redemption,9.3,142 min,Top Tier,82.000000,"['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469
1,2,The Godfather,9.2,175 min,Top Tier,100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,Top Tier,58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108
3,4,The Chaos Class,9.2,87 min,Top Tier,58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785
4,5,Daman,9.1,121 min,Top Tier,58.749625,"['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303


## Sorting Values

In [ ]:
movies_df.sort_values(by = ["Rating"], ascending=True)

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
9998,9999,The Pyramid,4.6,89 min,"Horror, Mystery, Thriller",24.000000,"['Grégory Levasseur', 'Ashley Hinshaw', 'James...","['Ashley Hinshaw', 'James Buckley', ""Denis O'H...",25483,2756333
9993,9994,Driven,4.6,116 min,"Action, Drama, Sport",29.000000,"['Renny Harlin', 'Sylvester Stallone', 'Kip Pa...","['Sylvester Stallone', 'Kip Pardue', 'Til Schw...",42226,32720065
9992,9993,Psycho,4.6,105 min,"Horror, Mystery, Thriller",47.000000,"['Gus Van Sant', 'Vince Vaughn', 'Anne Heche',...","['Vince Vaughn', 'Anne Heche', 'Julianne Moore...",49548,21456130
9994,9995,Yogi Bear,4.6,80 min,"Animation, Adventure, Comedy",35.000000,"['Eric Brevig', 'Dan Aykroyd', 'Justin Timberl...","['Dan Aykroyd', 'Justin Timberlake', 'Anna Far...",22762,100246011
9997,9998,Dark Crimes,4.6,92 min,"Crime, Drama, Mystery",24.000000,"['Alexandros Avranas', 'Jim Carrey', 'Marton C...","['Jim Carrey', 'Marton Csokas', 'Charlotte Gai...",19098,19098
...,...,...,...,...,...,...,...,...,...,...
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411


In [ ]:
movies_df.sort_values(by = ["Rating", "Votes"], ascending=[True, False])

,ID,title,Rating,Runtime,Genre,Metascore,Directors,Stars,Votes,Gross
9988,9989,Aliens vs. Predator: Requiem,4.6,94 min,"Action, Adventure, Horror",29.000000,"['Colin Strause', 'Greg Strause', 'Reiko Ayles...","['Greg Strause', 'Reiko Aylesworth', 'Steven P...",129456,41797066
9983,9984,Fifty Shades Darker,4.6,118 min,"Drama, Romance, Thriller",33.000000,"['James Foley', 'Dakota Johnson', 'Jamie Dorna...","['Dakota Johnson', 'Jamie Dornan', 'Eric Johns...",106578,114376885
9985,9986,Death Note,4.6,101 min,"Crime, Drama, Fantasy",43.000000,"['Adam Wingard', 'Nat Wolff', 'LaKeith Stanfie...","['Nat Wolff', 'LaKeith Stanfield', 'Margaret Q...",87627,87627
9984,9985,Superhero Movie,4.6,75 min,"Action, Comedy, Sci-Fi",33.000000,"['Craig Mazin', 'Drake Bell', 'Leslie Nielsen'...","['Drake Bell', 'Leslie Nielsen', 'Sara Paxton'...",72271,26638520
9996,9997,Paranormal Activity 4,4.6,88 min,"Horror, Mystery",40.000000,"['Henry Joost', 'Ariel Schulman', 'Stephen Dun...","['Ariel Schulman', 'Stephen Dunham', 'Katie Fe...",68430,53921300
...,...,...,...,...,...,...,...,...,...,...
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108


## Data Exploration

### The `.describe()` Method

Using `describe()` on an entire `DataFrame` we can get a summary of the distribution of *numeric* variables

In [ ]:
movies_df.describe()

,ID,Rating,Metascore,Votes,Gross
count,9999.000000,9999.000000,9999.000000,9.999000e+03,9.999000e+03
mean,5000.000000,6.698620,58.749625,9.127103e+04,2.874067e+07
std,2886.607005,0.852755,15.620576,1.684508e+05,5.971933e+07
min,1.000000,4.600000,7.000000,1.000100e+04,7.000000e+00
25%,2500.500000,6.100000,49.000000,1.678500e+04,3.654100e+04
50%,5000.000000,6.700000,58.749625,3.391200e+04,4.244155e+06
75%,7499.500000,7.300000,68.000000,8.965600e+04,3.306295e+07
max,9999.000000,9.300000,100.000000,2.752419e+06,9.366622e+08


It shows the non-null count, mean, standard deviation, min and max values, and the values of the quartiles of each numeric variable.

In the above, the summary of the `ID` column is irrelevant since it does not represent actual data.

The `Runtime` column contains the length of each movie in minutes, this should logically make the column `integer` type, however, each value is suffixed with the word `"min"`, therefore, the column is `string` type. We need to split the values of this column, take the first item after splitting and convert that to integer type.

In [ ]:
movies_df['Runtime'] = movies_df['Runtime'].str.split().str[0].astype(int)

In [ ]:
movies_df.drop('ID', axis = 1).describe()    # summary without the ID column

,Rating,Runtime,Metascore,Votes,Gross
count,9999.000000,9999.000000,9999.000000,9.999000e+03,9.999000e+03
mean,6.698620,110.479148,58.749625,9.127103e+04,2.874067e+07
std,0.852755,21.995584,15.620576,1.684508e+05,5.971933e+07
min,4.600000,45.000000,7.000000,1.000100e+04,7.000000e+00
25%,6.100000,96.000000,49.000000,1.678500e+04,3.654100e+04
50%,6.700000,106.000000,58.749625,3.391200e+04,4.244155e+06
75%,7.300000,120.000000,68.000000,8.965600e+04,3.306295e+07
max,9.300000,439.000000,100.000000,2.752419e+06,9.366622e+08


The `.describe()` method can also be run on individual non-numeric columns:

In [ ]:
movies_df['Genre'].describe()

count                       9999
unique                       428
top       Comedy, Drama, Romance
freq                         480
Name: Genre, dtype: object

It shows the count of the non-null values, the number of unique values, the value that appears the most and how many times it appears (`freq`).

### The `.value_counts()` method
For a categorical column, we can also run `.value_counts()`:

In [ ]:
movies_df.Genre.value_counts()
# here, movies_df.Genre is the same as movies_df.['Genre']

Genre
Comedy, Drama, Romance     480
Drama                      433
Comedy, Drama              408
Drama, Romance             336
Action, Crime, Drama       325
                          ... 
Horror, Music, Thriller      1
Action, Romance              1
Action, Fantasy, Sci-Fi      1
Action, Horror, War          1
Mystery, Sci-Fi              1
Name: count, Length: 428, dtype: int64

### The `.corr()` Method
Returns a correlation matrix

In [ ]:
movies_df.drop('ID', axis = 1).corr(numeric_only=True)

,Rating,Runtime,Metascore,Votes,Gross
Rating,1.000000,0.345162,0.623556,0.307727,0.064945
Runtime,0.345162,1.000000,0.181856,0.197374,0.124496
Metascore,0.623556,0.181856,1.000000,0.202101,0.029370
Votes,0.307727,0.197374,0.202101,1.000000,0.632156
Gross,0.064945,0.124496,0.029370,0.632156,1.000000


## Computing New Columns

In [ ]:
# creating a column depticcting gross revenue per minute of runtime
movies_df['revenue_per_minute'] = movies_df['Gross'] / movies_df['Runtime']
movies_df['revenue_per_minute'].head()

0    199587.809859
1    771236.634286
2        74.874074
3       480.287356
4       109.942149
Name: revenue_per_minute, dtype: float64

### The `.transform()` Method

This method applies a function to an axis (either column or row) and returns a `DataFrame` of the same shape.

We can mention `axis=1` for applying the function across the columns (row-wise), meaning that with `axis=1`, the function will get individual rows as input. Otherwise, it is column-wise.

For example, if we wanted to standardize the rating column:

In [ ]:
movies_df['rating_standardized'] = movies_df['Rating'].transform(lambda x: (x-x.mean())/x.std())

In [ ]:
movies_df['rating_standardized'].head()

0    3.050559
1    2.933292
2    2.933292
3    2.933292
4    2.816025
Name: rating_standardized, dtype: float64

In [ ]:
movies_df['rating_standardized'].mean()

np.float64(-9.095856603389622e-17)

In [ ]:
movies_df['rating_standardized'].std()

np.float64(1.0)

### The `.assign()` method

Assigns new columns to a `DataFrame`. Returns a new object with all original columns in addition to new ones. Existing columns that are re-assigned will be overwritten.

In [ ]:
movies_df = movies_df.assign(
    rating_out_of_1 = movies_df['Rating'] / 10,
    revenue_std = lambda df: (df['Gross'] - df['Gross'].mean())/df['Gross'].std()
)

movies_df[['rating_out_of_1', 'revenue_std']].head()

,rating_out_of_1,revenue_std
0,0.93,-0.006685
1,0.92,1.778750
2,0.92,-0.481093
3,0.92,-0.480563
4,0.91,-0.481040


### Discretization and Binning

Sometimes we need to convert numeric/continuous columns to categorical/discrete columns.

#### Using `pd.cut()`

This functions "cuts" a numeric/continuous variable into several groups making it discrete or categorical.

```
pandas.cut(x, bins, right=True, labels=None, retbins=False, precision=3, include_lowest=False, duplicates='raise', ordered=True)
```

Parameters:
- x: array-like (series/column). The input array to be binned. Must be 1-dimensional.

- bins: int, sequence of scalars, or IntervalIndex. The criteria to bin by.
    - int : Defines the number of equal-width bins in the range of x. The range of x is extended by .1% on each side to include the minimum and maximum values of x.
    - sequence of scalars : Defines the bin edges allowing for non-uniform width. No extension of the range of x is done.
    - IntervalIndex : Defines the exact bins to be used. Note that IntervalIndex for bins must be non-overlapping.
- right: bool, default=True. Indicates whether bins includes the rightmost edge or not. If right == True (the default), then the bins [1, 2, 3, 4] indicate (1,2], (2,3], (3,4]. This argument is ignored when bins is an IntervalIndex.
- labels: array or False, default=None. Specifies the labels for the returned bins. Must be the same length as the resulting bins.
- include_lowest: bool, default=False. Whether the first interval should be left-inclusive or not.
- ordered: bool, default=True. Whether the labels are ordered or not. Applies to returned types Categorical and Series (with Categorical dtype). If True, the resulting categorical will be ordered. If False, the resulting categorical will be unordered (labels must be provided).



In [ ]:
rating_cat = pd.cut(movies_df['Rating'], 3)
rating_cat.value_counts()

Rating
(6.167, 7.733]    6255
(4.595, 6.167]    2650
(7.733, 9.3]      1094
Name: count, dtype: int64

Can mention `right = False` to include lower limits and not upper limits:

In [ ]:
rating_cat2 = pd.cut(movies_df['Rating'], 3, right = False)
rating_cat2.value_counts()

Rating
[6.167, 7.733)    6255
[4.6, 6.167)      2650
[7.733, 9.305)    1094
Name: count, dtype: int64

In [ ]:
rating_cat3 = pd.cut(movies_df['Rating'], 3, include_lowest=True)
rating_cat3.value_counts()

Rating
(6.167, 7.733]                6255
(4.593999999999999, 6.167]    2650
(7.733, 9.3]                  1094
Name: count, dtype: int64

In [ ]:
rating_cat4 = pd.cut(movies_df.Rating, [0, 5, 8, 10], right = True)
rating_cat4.value_counts()

Rating
(5, 8]     9204
(8, 10]     496
(0, 5]      299
Name: count, dtype: int64

Can also provide a `list` of labels for each category using the `labels` argument:

In [ ]:
rating_cat5 = pd.cut(movies_df.Rating, [0, 5, 8, 10], labels = ['<=5', '5-8', '>8'])
rating_cat5.value_counts()

Rating
5-8    9204
>8      496
<=5     299
Name: count, dtype: int64

#### `pd.qcut()`

Quantile-based discretization function. Uses sample quantiles to bin.

```
Parameters
----------
x : 1d ndarray or Series
q : int or list-like of float
    int: Number of quantiles. 10 for deciles, 4 for quartiles, etc.
    list: Alternately array of quantiles, e.g. [0, .25, .5, .75, 1.] for quartiles.
labels : array or False, default None
    Used as labels for the resulting bins. Must be of the same length as
    the resulting bins. If False, return only integer indicators of the
    bins. If True, raises an error.
retbins : bool, optional
    Whether to return the (bins, labels) or not. Can be useful if bins
    is given as a scalar.
```

In [ ]:
pd.qcut(x = movies_df.Rating, q = 3).value_counts()

Rating
(4.598999999999999, 6.3]    3441
(6.3, 7.1]                  3309
(7.1, 9.3]                  3249
Name: count, dtype: int64

In [ ]:
pd.qcut(x = movies_df.Rating, q = [0, 0.25, 0.5, 0.75, 1]).value_counts()

Rating
(4.598999999999999, 6.1]    2650
(6.1, 6.7]                  2476
(6.7, 7.3]                  2468
(7.3, 9.3]                  2405
Name: count, dtype: int64

### Aggregation

Aggregation functions are used to obtain summary information on columns of a dataframe using the `.agg()` method

#### Method 1: providing a callable function function to `.agg()`

In [ ]:
import numpy as np

In [ ]:
movies_df['Gross'].mean()

np.float64(28740668.220622063)

In [ ]:
movies_df[["Gross", "Rating"]].agg("mean")

Gross     2.874067e+07
Rating    6.698620e+00
dtype: float64

Can also pass a lambda function:

In [ ]:
movies_df[["Rating", "Gross"]].agg(lambda x: np.mean(x))

Rating    6.698620e+00
Gross     2.874067e+07
dtype: float64

In [ ]:
def get_summaries(x):
    return pd.Series(
        [x.mean(), x.std()],
        index = ["mean", "stdev"]
    )

In [ ]:
movies_df[["Rating", "Gross"]].agg(get_summaries)

,Rating,Gross
mean,6.698620,2.874067e+07
stdev,0.852755,5.971933e+07


#### Method 2: Selecting the columns, then providing the aggregation function name as strings to `.agg()` as a list of string:

In [ ]:
movies_df[["Gross", "Rating"]].agg(["max", "min", "mean"])

,Gross,Rating
max,9.366622e+08,9.30000
min,7.000000e+00,4.60000
mean,2.874067e+07,6.69862


#### Method 3: Passing tuples of column name and aggregate function name as string

In [ ]:
movies_df.agg(max_value = ("Gross", "max"), min_value = ("Gross", "min"), mean_rating = ("Rating", "mean"))

,Gross,Rating
max_value,936662225.0,NaN
min_value,7.0,NaN
mean_rating,NaN,6.69862


#### Method 4: Passing a dictionary with keys as column names on which to perform the aggregation, and lists of aggregating function names as strings as the values to the corresponding keys of the dictionary

In [ ]:
movies_df.agg({
    "Rating" : ["mean", "min", "max"],
    "Gross" : ["min", "max", "std"]
})

,Rating,Gross
mean,6.69862,NaN
min,4.60000,7.000000e+00
max,9.30000,9.366622e+08
std,NaN,5.971933e+07


Below is a non-comprehensive list of `pandas` aggregation functions:

| Function      | Description |
|---------------|-------------|
| sum           | Adds up the values. |
| mean          | Computes the average of the values. |
| median        | Returns the median value. |
| min           | Gives the smallest value. |
| max           | Gives the largest value. |
| count         | Counts non-missing values. |
| size          | Counts all rows, including missing values. |
| nunique       | Counts distinct values. |
| std           | Computes standard deviation. |
| var           | Computes variance. |
| prod          | Multiplies all values. |
| first         | Returns the first non-missing value. |
| last          | Returns the last non-missing value. |
| idxmin        | Returns the index of the minimum value. |
| idxmax        | Returns the index of the maximum value. |
| quantile      | Returns a specified quantile. |
| mad           | Computes mean absolute deviation. |
| sem           | Computes standard error of the mean. |
| skew          | Measures distribution skewness. |
| kurt          | Measures distribution kurtosis. |
| corr          | Computes correlation with another series. |
| cov           | Computes covariance with another series. |
| mode          | Returns the mode(s) of the values. |
| any           | Tests if any element is True. |
| all           | Tests if all elements are True. |

Apart from the above, some `numpy` universal functions (`unfuncs`) can also be used.

### Applying Functions and Mapping


A frequent operation while analyzing a data is to apply a function on one-dimensional arrays to each column or each row. It can be done with the `.apply()` method of a `DataFrame`.

The `.apply()` method takes a function as its argument as well as the axis *across* which the function is applied. The function can be built-in, from a package or a lambda.

For example, we can define a function that returns the range and the 90th percentile of a variable:

In [ ]:
def mysummary(x):
    return [x.max() - x.min(), x.quantile(0.90)]

If we want to find the range and the median of say, rating and gross revenue:

In [ ]:

movies_df[['Rating', 'Gross']].apply(mysummary, axis=0)

,Rating,Gross
0,4.7,936662218.0
1,7.8,80572126.0


We can see that the two summary statistics returned from our defined function, namely, range and 90th percentile are indexed as 0 and 1, respectively in the output.

While outputting multiple summary statistics, it is desirable to index those statistics with appropriate labels.

We can provide appropriate labels to our outputs by returning a Series instead of a list as follows:

In [ ]:
def mysummary(x):
    return pd.Series([x.max() - x.min(), x.quantile(0.90)],
                    index = ['range', '90th Percentile'])
movies_df[['Runtime', 'Gross']].apply(mysummary, axis=0)

,Runtime,Gross
range,394.0,936662218.0
90th Percentile,138.0,80572126.0


We can also invoke lambda functions inside the `.apply()` method instead of defining a function using the `def` keyword:

In [ ]:
movies_df[['Runtime', 'Gross']].apply(
    lambda x: pd.Series(
        [x.max() - x.min(), x.quantile(0.90)],
        index = ['range', '90th Percentile']
    )
)

,Runtime,Gross
range,394.0,936662218.0
90th Percentile,138.0,80572126.0


Using apply() will be much faster than iterating manually over rows because pandas is utilizing vectorization.

### Grouped aggregation

When exploring and analyzing data, we often need to perform grouped analysis. For example, we may want to find the mean rating across the various genre of movies in the dataset.

Since many of the movies belong to muultiple genre, there are too many combination of genres. So, we create a new column which keeps only the first genre of each movie:

In [ ]:
movies_df['genre_one'] = movies_df['Genre'].str.split(',').str[0]

In [ ]:
movies_df['genre_one'].value_counts()

genre_one
Comedy       2546
Action       2355
Drama        2033
Crime         808
Biography     581
Adventure     544
Horror        487
Animation     483
Fantasy        60
Mystery        41
Thriller       22
Sci-Fi          9
Film-Noir       8
Western         7
Romance         6
Family          4
Musical         2
Music           1
War             1
History         1
Name: count, dtype: int64

In [ ]:
movies_df.groupby('genre_one')['Rating'].agg(('mean', 'std')).sort_values('mean')

,mean,std
genre_one,,
Horror,6.057084,0.761298
Fantasy,6.073333,0.851891
Thriller,6.186364,0.749155
History,6.200000,NaN
Sci-Fi,6.455556,0.745170
Action,6.508662,0.847091
Family,6.575000,0.865544
Comedy,6.579301,0.811387
Adventure,6.697059,0.888265


In [ ]:
# we can also obtain the mean of all numeric columns:
(
    movies_df
    .drop('ID', axis = 'columns')
    .groupby('genre_one')
    .mean(numeric_only=True)
)

,Rating,Runtime,Metascore,Votes,Gross,revenue_per_minute,rating_standardized,rating_out_of_1,revenue_std
genre_one,,,,,,,,,
Action,6.508662,116.259873,54.127302,133757.068790,4.655420e+07,384735.011586,-0.222757,0.650866,0.298288
Adventure,6.697059,111.858456,58.425024,105576.055147,3.993996e+07,345750.750802,-0.001831,0.669706,0.187532
Animation,6.939545,93.053830,63.501485,119680.146998,6.856429e+07,731599.455863,0.282525,0.693954,0.666846
Biography,7.103270,122.320138,64.098065,89606.306368,2.091005e+07,165283.970199,0.474521,0.710327,-0.131124
Comedy,6.579301,104.664965,57.012691,61693.785546,2.284620e+07,219980.202558,-0.139922,0.657930,-0.098703
Crime,6.960520,112.345297,62.282406,92423.574257,1.343133e+07,112632.104299,0.307122,0.696052,-0.256355
Drama,6.963355,114.678308,63.925404,73330.230694,1.452171e+07,120406.028478,0.310446,0.696335,-0.238096
Family,6.575000,106.000000,56.937406,67858.500000,5.568126e+06,55112.250952,-0.144965,0.657500,-0.388024
Fantasy,6.073333,98.616667,52.454073,66293.233333,2.183054e+07,199929.011013,-0.733254,0.607333,-0.115710


### More `.groupby` examples

**Note on the dataset used below**

The examples in this subsection switch away from `movies_df` to a small, hand built state population dataset. This is intentional: the point here is to isolate the mechanics of `.groupby()` with weighted averages, custom aggregation via `.apply()`, and `.transform()`, without the noise of a large real dataset. We return to `movies_df` in the next section on joining dataframes.

#### Groupby the summarize/aggregate

In [ ]:
data = {
    "state" : ["Ohio"]*3 + ["Nevada"]*3,
    "year" : [2000, 2001, 2002, 2001, 2002, 2003],
    "popn" : [1.5, 1.7, 3.6, 2.4, 2.9, 3.2],
    "wt" : [1, 2, 3] * 2    # weights for findingg weighted average
}
df = pd.DataFrame(data)
df

,state,year,popn,wt
0,Ohio,2000,1.5,1
1,Ohio,2001,1.7,2
2,Ohio,2002,3.6,3
3,Nevada,2001,2.4,1
4,Nevada,2002,2.9,2
5,Nevada,2003,3.2,3


In [ ]:
df.groupby("state").agg(avg_pop = ("popn", "mean"))    # returns a dataframe

,avg_pop
state,
Nevada,2.833333
Ohio,2.266667


In [ ]:
# finding weighted average
df.groupby("state").apply(lambda g: sum(g.popn * g.wt) / sum(g.wt))    # returns a Series

state
Nevada    2.966667
Ohio      2.616667
dtype: float64

In [ ]:
# wrapping the result in a Series makes the whole thing a dataframe
df.groupby("state").apply(
    lambda g: pd.Series(
        {"wt_avg" : sum(g.popn * g.wt) / sum(g.wt)}
    )
)

,wt_avg
state,
Nevada,2.966667
Ohio,2.616667


In [ ]:
# this returns a dataframe
agdf = (
    df.groupby(df.state).apply(
        lambda x: pd.Series(
            {
                "avg" : np.mean(x.popn),
                "wt_avg" : sum(x.popn * x.wt)/sum(x.wt),
                "wt_sum" : sum(x.popn * x.wt)
            }
            )
        )
    )

agdf

,avg,wt_avg,wt_sum
state,,,
Nevada,2.833333,2.966667,17.8
Ohio,2.266667,2.616667,15.7


#### Groupby then mutate

In [ ]:
df.join(agdf, on = "state")

,state,year,popn,wt,avg,wt_avg,wt_sum
0,Ohio,2000,1.5,1,2.266667,2.616667,15.7
1,Ohio,2001,1.7,2,2.266667,2.616667,15.7
2,Ohio,2002,3.6,3,2.266667,2.616667,15.7
3,Nevada,2001,2.4,1,2.833333,2.966667,17.8
4,Nevada,2002,2.9,2,2.833333,2.966667,17.8
5,Nevada,2003,3.2,3,2.833333,2.966667,17.8


In [ ]:
df.groupby(df.state)["popn"].transform(lambda x: x.mean())

0    2.266667
1    2.266667
2    2.266667
3    2.833333
4    2.833333
5    2.833333
Name: popn, dtype: float64

In [ ]:
df["avg_pop2"] = df.groupby(df.state)["popn"].transform(lambda x: x.mean())
df

,state,year,popn,wt,avg_pop2
0,Ohio,2000,1.5,1,2.266667
1,Ohio,2001,1.7,2,2.266667
2,Ohio,2002,3.6,3,2.266667
3,Nevada,2001,2.4,1,2.833333
4,Nevada,2002,2.9,2,2.833333
5,Nevada,2003,3.2,3,2.833333


In [ ]:
(
    df
    .groupby("state", group_keys = False)
    .apply(
        lambda grp: grp.assign(
            wt_avg2 = lambda x: sum(x.popn * x.wt) / sum(x.wt)
        ),
    )
)

,year,popn,wt,avg_pop2,wt_avg2
0,2000,1.5,1,2.266667,2.616667
1,2001,1.7,2,2.266667,2.616667
2,2002,3.6,3,2.266667,2.616667
3,2001,2.4,1,2.833333,2.966667
4,2002,2.9,2,2.833333,2.966667
5,2003,3.2,3,2.833333,2.966667


## Joining dataframes

`pd.merge()` (or the equivalent `.merge()` method) combines two `DataFrame`s based on the values in one or more shared key columns, similar to a SQL join.

Key parameters:

- `on`: column name (or list of names) to join on, when both frames use the same column name for the key.
- `left_on` / `right_on`: used instead of `on` when the key column has different names in each frame.
- `how`: the join type.
    - `inner` (default): keep only keys present in both frames.
    - `left`: keep all keys from the left frame, filling missing right-side values with `NaN`.
    - `right`: keep all keys from the right frame, filling missing left-side values with `NaN`.
    - `outer`: keep all keys from either frame, filling missing values with `NaN` on whichever side lacks the key.
- `suffixes`: a tuple of strings appended to overlapping non-key column names, so both can be kept without a name clash.
- `indicator`: if `True`, adds a `_merge` column labelling each row as `left_only`, `right_only`, or `both`.

In [ ]:
df1 = movies_df[['ID', 'title', 'Runtime']]
df2 = movies_df[['ID', 'title', 'Rating']]

In [ ]:
dfjoin = df1.merge(df2, on = "ID", how = "inner")
dfjoin.info()

<class 'pandas.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       9999 non-null   int64  
 1   title_x  9999 non-null   str    
 2   Runtime  9999 non-null   int64  
 3   title_y  9999 non-null   str    
 4   Rating   9999 non-null   float64
dtypes: float64(1), int64(2), str(2)
memory usage: 390.7 KB


## Exercises

Use `movies_df` (after loading and cleaning it as done in this notebook) unless a question says otherwise.

1. Create a `Series` of the five highest `Gross` values in the dataset, with the corresponding `title` as the index. Do not use `.sort_values()`; look for an alternative method that returns the top `n` values directly.

2. Using `.iloc`, select the last five rows of `movies_df` and only the `title` and `Rating` columns.

3. Select all movies where `Genre` is either `'Drama'` or `'Crime'` and `Rating` is at least `8.5`. How many rows does this return?

4. Create a new column `decade` that bins each movie's release information into 10 year buckets using `pd.cut()`. (If a release year column is not already present, first inspect `movies_df.columns` to find the correct one.)

5. Using `.groupby()` on `genre_one` (created earlier in this notebook), find the mean and standard deviation of `Runtime` for each genre, sorted from longest to shortest average runtime.

6. Write a function that takes a `Series` and returns a `Series` indexed by `'min'`, `'max'`, and `'range'`. Apply it to the `Votes` column using `.agg()`.